Add flag for newline before toplevel statements

Daxten opened the pull request:
Hi, thanks for the help in Gitter!
Here is my first try to add a  flag for adding newlines before toplevel statements

olafurpg commented on core/src/main/scala/org/scalafmt/config/Newlines.scala:
```diff
@@ -30,5 +30,6 @@ case class Newlines(
     penalizeSingleSelectMultiArgList: Boolean = true,
-    alwaysBeforeCurlyBraceLambdaParams: Boolean = false
+    alwaysBeforeCurlyBraceLambdaParams: Boolean = false,
+    beforeTopLevelStatements: Boolean = false
```
`alwaysBeforeTopLevelStatements`?

pjrt replied:
Could you add some doc for this variable?

pjrt commented:
I can see why we want this as a flag now (not to break everything) but I think this should eventually be the default, along with #732 .

Maybe for v1? #730

olafurpg commented on core/src/main/scala/org/scalafmt/util/TokenOps.scala:
```diff
@@ -52,10 +53,14 @@ object TokenOps {
+    } || {
+      style.newlines.beforeTopLevelStatements &&
+        tok.between.count(_.is[KwNew]) < 2 &&
+        (tok.right.is[KwObject] || tok.right.is[KwDef] || tok.right.is[KwDef] || tok.right.is[KwCase] || tok.right.is[KwPackage])
```
FYI, this won't cover modifiers like `final class` or `sealed trait`. Depending on your use-case, you might want to support that. There's a `owners: Token => Tree` available in Router that you can use to get the tree parent of the token. Note that the owner of modifiers like `lazy` is the Mod tree while the owner of `class` for exampl is Defn.Class.

btw, KwDef appears twice.

pjrt replied:
I would think modified top levels (`final`, `sealed`) would also need the newline. 

Also, maybe refactor that into a "isTopLevelStatment" function in FormatOps?

olafurpg replied:
There is a `startsStatement: Set[Token]` iirc, if that was a Map[Token, Tree] then we could more easily support that

Daxten commented:
So, I've added a failing testcase and added everything which is needed (I think?) but I can't find a solution which works using owners, can someone give me a hand?

Daxten commented:
Thanks @olafurpg wasn't so hard after all!

olafurpg commented on core/src/main/scala/org/scalafmt/util/TokenOps.scala:
```diff
@@ -52,13 +54,26 @@ object TokenOps {
+        tok.is[KwCase] ||
+        tok.is[KwPackage] ||
+        tok.is[KwClass] ||
+        owners(tok).parent.exists(_.is[Defn.Class]) ||
```
Can we combine these two lines? I would suggest to replace the Kw* matches with matches against the owner tree.

Daxten replied:
the KwClass is actualy redundant yes, I removed that. Do you mean that also the rest should be changed?

olafurpg commented on core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat:
```diff
@@ -55,9 +61,17 @@ package a {
 
+      sealed trait T1
+
+      final class FC1
```
These blank lines would be preserved with default

Daxten replied:
Yes, that's correct, since this Test just tests if the new formatter doesn't add any new lines in this case

olafurpg commented on readme/Configuration.scalatex:
```diff
@@ -204,6 +204,35 @@
+  @sect{newlines.alwaysBeforeTopLevelStatements}
+    Default: default.newlines.alwaysBeforeTopLevelStatements
+
+    @h1.scala
```
This is failing the CI, I think it should be `hl.scala`.

olafurpg commented:
Thank you for this contribution @Daxten!

Daxten commented:
maybe add a second flag to disable it for objects ( / packages ) which only contain class (/ case / traits) definitions? Dunno if this gets too spaghetti
